api/stream-relay/index.js

```diff
@@ -1007,79 +1113,161 @@ export const wrap = dgram => {
+
+            const pid = packet.packetId.toString('hex')
+            debug(this.peerId, `-> SYNC SEND PACKET (type=data, packetId=${pid.slice(0, 8)}, to=${address}:${port})`)
+
+            this.send(await Packet.encode(packet), port, address)
           }
-        } else if (remote.buckets[i] !== local.buckets[i]) {
+        } else {
           //
           // need more details about what exactly isn't synce'd
           //
-          const nextLevel = await this.cache.summarize(local.prefix + i.toString(16))
-          const encoded = Cache.encodeSummary(nextLevel)
+          const nextLevel = await this.cache.summarize(local.prefix + i.toString(16), this.cachePredicate)
           const data = await Packet.encode(new PacketSync({
-            message: encoded
+            message: Cache.encodeSummary(nextLevel)
           }))
-          this._setTimeout(() => this.send(data, port, address), 2048)
+          this.send(data, port, address)
         }
       }
     }
 
     /**
      * Received a Query Packet
+     *
+     * a -> b -> c -> (d) -> c -> b -> a
+     *
      * @return {undefined}
-     * @ignore
+     * @example
+     *
+     * ```js
+     * peer.onQuery = (packet) => {
+     *   //
+     *   // read a database or something
+     *   //
+     *   return {
+     *     message: Buffer.from('hello'),
+     *     publicKey: '',
+     *     privateKey: ''
+     *   }
+     * }
+     * ```
      */
     async _onQuery (packet, port, address) {
-      this.metrics[packet.type]++
+      this.metrics.i[packet.type]++
 
-      debug(this.peerId, '<- QUERY', port, address)
-      if (this.controlPackets.has(packet.packetId)) return
-      this.controlPackets.set(packet.packetId, 1)
+      const pid = packet.packetId.toString('hex')
+      if (this.gate.has(pid)) return
+      this.gate.set(pid, 1)
 
-      if (this.onEvaluateQuery) await this.onEvaluateQuery(packet, port, address)
+      const queryTimestamp = parseInt(Buffer.from(packet.usr1).toString(), 10)
+      const queryId = Buffer.from(packet.usr3).toString('hex')
+      const queryType = parseInt(Buffer.from(packet.usr4).toString(), 10)
 
-      if (!Array.isArray(packet.message.history)) {
-        packet.message.history = []
-      }
+      // if the timestamp in usr1 is older than now - 2s, bail
+      if (queryTimestamp < (Date.now() - 2048)) return
+
+      const type = queryType === 1 ? 'question' : 'answer'
+      debug(this.peerId, `<- QUERY (type=${type}, from=${address}:${port}, packet=${pid.slice(0, 8)})`)
 
-      packet.message.history.push({ address: this.address, port: this.port })
-      if (packet.message.history.length >= 16) packet.message.history.shift()
+      let rinfo = { port, address }
+
+      //
+      // receiving an answer
+      //
+      if (this.returnRoutes.has(queryId)) {
+        rinfo = this.returnRoutes.get(queryId)
+
+        let p = packet.copy()
+        if (p.index > -1) p = await this.cache.compose(p)
+
+        if (p && p.index === -1) {
```

Comment:
hmmm is `p &&` necessary here? it will throw on line 1220 if it's not truthy. or maybe it should be in 1220 there as well?

Reply:
Switch to `p?.index`?